controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java

```diff
@@ -0,0 +1,158 @@
+        this.controller = controller;
+        this.clock = controller.clock();
+    }
+
+    /** Returns a list of jobs to trigger after the given job */
+    public List<JobType> nextAfter(JobType job, Application application) {
+        // Always trigger system test after component as deployment spec might not be available yet (e.g. if this is a
+        // new application with no previous deployments)
+        if (job == JobType.component) {
+            return Collections.singletonList(JobType.systemTest);
+        }
+
+        // At this point we have deployed to system test, so deployment spec is available
+        List<DeploymentSpec.Step> deploymentSteps = deploymentSteps(application);
+        Optional<DeploymentSpec.Step> currentStep = fromJob(job, application);
+        if (!currentStep.isPresent()) {
+            return Collections.emptyList();
+        }
+
+        // If this is the last deployment step there's nothing more to trigger
+        if (isLast(job, application)) {
+            return Collections.emptyList();
+        }
+
+        // Postpone next job if delay has not passed yet
+        Duration delay = delayAfter(currentStep.get(), application);
+        if (postponeDeployment(delay, job, application)) {
+            log.info(String.format("Delaying next job after %s of %s by %s", job, application, delay));
+            return Collections.emptyList();
+        }
+
+        DeploymentSpec.Step nextStep = deploymentSteps.get(deploymentSteps.indexOf(currentStep.get()) + 1);
+        if (nextStep instanceof DeploymentSpec.DeclaredZone) {
+            return Collections.singletonList(toJob((DeploymentSpec.DeclaredZone) nextStep));
+        } else if (nextStep instanceof DeploymentSpec.ParallelZones) {
+            return ((DeploymentSpec.ParallelZones) nextStep).zones().stream()
+                    .map(this::toJob)
+                    .collect(collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
+        } else {
+            throw new IllegalStateException("Unexpected step type: " + nextStep.getClass());
+        }
+    }
+
+    /** Returns whether the given job is first in a deployment */
+    public boolean isFirst(JobType job) {
+        return job == JobType.component;
+    }
+
+    /** Returns whether the given job is last in a deployment */
+    public boolean isLast(JobType job, Application application) {
+        List<DeploymentSpec.Step> deploymentSteps = deploymentSteps(application);
+        if (deploymentSteps.isEmpty()) { // Deployment spec not yet available
+            return false;
+        }
+        DeploymentSpec.Step lastStep = deploymentSteps.get(deploymentSteps.size() - 1);
+        return fromJob(job, application).get().equals(lastStep);
+    }
+
+    /** Returns jobs for given deployment spec, in the order they are declared */
+    public List<JobType> jobsFrom(DeploymentSpec deploymentSpec) {
+        if (deploymentSpec.steps().isEmpty()) {
```

Comment:
I think you should remove this if block. DeploymentSpec.steps() should be authorative on which steps to perform.

Reply:
Yes, it stems from the fact that a test breaks encapsulation by calling into DeploymentTrigger directly. I'll fix the test and remove the empty check.